ompi/runtime/ompi_spc.c

```diff
@@ -0,0 +1,510 @@
+                                     ompi_spc_get_count, NULL, ompi_spc_notify, NULL);
+
+        /* Initialize the mpi_t_indices array with the MPI_T indices.
+         * The array index indicates the SPC index, while the value indicates
+         * the MPI_T index.
+         */
+        if(ret != OPAL_ERROR) {
+            mpi_t_indices[i] = ret;
+
+            if(mpi_t_offset == -1) {
+                mpi_t_offset = ret;
+                prev = ret;
+            } else if(ret != prev + 1) {
+                mpi_t_disabled = 1;
+                fprintf(stderr, "There was an error registering SPCs as MPI_T pvars.  SPCs will be disabled for MPI_T.\n");
+                break;
+            } else {
+                prev = ret;
+            }
+        } else {
+            mpi_t_indices[i] = -1;
+            mpi_t_disabled = 1;
+            fprintf(stderr, "There was an error registering SPCs as MPI_T pvars.  SPCs will be disabled for MPI_T.\n");
+            break;
+        }
+    }
+}
+
+/* Frees any dynamically alocated OMPI SPC data structures */
+void ompi_spc_fini()
+{
+#if SPC_ENABLE == 1
+    if(!ompi_mpi_spc_dump_enabled)
+        goto skip_dump;
+
+    int i, j, rank, world_size, offset, err;
+    long long *recv_buffer, *send_buffer;
+
+    ompi_communicator_t *comm = &ompi_mpi_comm_world.comm;
+
+    rank = ompi_comm_rank(comm);
+    world_size = ompi_comm_size(comm);
+
+    /* Convert from cycles to usecs before sending */
+    for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+        if(timer_event[i])
+            SPC_CYCLES_TO_USECS(&events[i].value);
+    }
+
+    /* Aggregate all of the information on rank 0 using MPI_Gather on MPI_COMM_WORLD */
+    if(rank == 0) {
+        send_buffer = (long long*)malloc(OMPI_NUM_COUNTERS * sizeof(long long));
+        recv_buffer = (long long*)malloc(world_size * OMPI_NUM_COUNTERS * sizeof(long long));
+        for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+            send_buffer[i] = events[i].value;
+        }
+        err = comm->c_coll->coll_gather(send_buffer, OMPI_NUM_COUNTERS, MPI_LONG_LONG,
+                                        recv_buffer, OMPI_NUM_COUNTERS, MPI_LONG_LONG,
+                                        0, comm,
+                                        comm->c_coll->coll_gather_module);
+    } else {
+        send_buffer = (long long*)malloc(OMPI_NUM_COUNTERS * sizeof(long long));
+        for(i = 0; i < OMPI_NUM_COUNTERS; i++) {
+            send_buffer[i] = events[i].value;
+        }
+        err = comm->c_coll->coll_gather(send_buffer, OMPI_NUM_COUNTERS, MPI_LONG_LONG,
+                                        recv_buffer, OMPI_NUM_COUNTERS, MPI_LONG_LONG,
+                                        0, comm,
+                                        comm->c_coll->coll_gather_module);
+    }
+
+    /* Once rank 0 has all of the information, print the aggregated counter values for each rank in order */
+    if(rank == 0) {
+        fprintf(stdout, "OMPI Software Counters:\n");
```

Comment:
Should we use `fprintf` or `opal_output` here?

Also, we use the name "Open MPI" publicly; we rarely use the name `OMPI` in help/output messages.